### PR TITLE
VIDEO-7161 + VIDEO-7162 Simulcast API implementation + rollback support 

### PR DESCRIPTION
**Pull request by makarandp0**
This change implements the proposed API for adaptive simulcast.

Now  `connect` will accept a new value 'auto' for preferredVideoCodecs. This new values would enable adaptive simulcast  and request for 'publisher_hints' msp channel in initial connect message.

To handle the case for P2P room (where we do not want to enable simulcast) SDK will check for availability of publisher_hints msp in `connected` message, and if not available then rollback the offer that was sent with simulcast enabled.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Review by manjeshbhargav — CHANGES_REQUESTED** [review]

File: lib/connect.js [hunk 1]
```diff
@@ -217,8 +217,12 @@ function connect(token, options) {
   // Additionally, the options that are no longer needed will be removed.
   deprecateOptions(options, log, deprecatedConnectOptionsProps);
 
+  if (options.preferredVideoCodecs === 'auto') {
+    // NOTE(mpatwardhan): enable adaptiveSimulcast.
+    options.preferredVideoCodecs = [{ codec: 'VP8', simulcast: true, adaptiveSimulcast: true }];
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/connect.js:222]: Good point, so far its decided to that this not part of public API.  we could maintain this settings separately but it really belongs here, Once we move to typescript  we could cast this type as 
```
export interface VP8CodecSettings extends VideoCodecSettings {
  codec: 'VP8';
  simulcast?: boolean;
}

export interface InternalVP8CodecSettings extends VP8CodecSettings {
  adaptiveSimulcast?: boolean;
}

```

File: lib/signaling/v2/peerconnection.js [hunk 2]
```diff
@@ -428,6 +423,28 @@ class PeerConnectionV2 extends StateMachine {
     return `[PeerConnectionV2 #${this._instanceId}: ${this.id}]`;
   }
 
+  setEffectiveAdaptiveSimulcast(effectiveAdaptiveSimulcast) {
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js:426]: This is called from Room, not sure making it '_' gives wrong impression that its private/protected.

**Review by manjeshbhargav — APPROVED** [review]
LGTM. However, please create a ticket to add a CHANGELOG.md entry for this feature.

**Comment by KaelWD** [comment]
Does this resolve #1398? We want to disable simulcast in group rooms with only two participants as well, not just P2P rooms.

**Pull request merged**